### PR TITLE
executor: refine hash join v2 for spill

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: ref #55153

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/join/hash_join_base.go [hunk 1]
```diff
@@ -218,15 +238,44 @@ type buildWorkerBase struct {
...
+
 func (w *buildWorkerBase) fetchBuildSideRows(ctx context.Context, hashJoinCtx *hashJoinCtxBase, chkCh chan<- *chunk.Chunk, errCh chan<- error, doneCh <-chan struct{}) {
-	defer close(chkCh)
-	var err error
+	w.fetchBuildSideRowsImpl(ctx, hashJoinCtx, nil, chkCh, errCh, doneCh)
+}
+
+// fetchBuildSideRowsImpl fetches all rows from build side executor, and append them
+// to e.buildSideResult.
+func (w *buildWorkerBase) fetchBuildSideRowsImpl(ctx context.Context, hashJoinCtx *hashJoinCtxBase, fetcherAndWorkerSyncer *sync.WaitGroup, chkCh chan<- *chunk.Chunk, errCh chan<- error, doneCh <-chan struct{}) {
```
Comment (windtalker) [inline pkg/executor/join/hash_join_base.go:255]: why not add `fetcherAndWorkerSyncer` as an argument of `fetchBuildSideRows` and let the caller to pass this argument

File: pkg/executor/join/hash_join_base.go [hunk 2]
```diff
@@ -218,15 +234,44 @@ type buildWorkerBase struct {
...
+	var globalErr error
+	defer func() {
+		defer close(chkCh)
```
Comment (windtalker) [inline pkg/executor/join/hash_join_base.go:258]: You can define mulitple defers in a function, no need to use nested defer. For this specific case, you can move `defer close(chkCh)` to the first line of this function.

File: pkg/executor/join/hash_join_base.go [hunk 3]
```diff
@@ -239,36 +284,51 @@ func (w *buildWorkerBase) fetchBuildSideRows(ctx context.Context, hashJoinCtx *h
...
 		failpoint.Inject("ConsumeRandomPanic", nil)
+
 		if chk.NumRows() == 0 {
 			return
 		}
+
+		syncerAdd(fetcherAndWorkerSyncer)
 		select {
 		case <-doneCh:
+			syncerDone(fetcherAndWorkerSyncer)
```
Comment (windtalker) [inline pkg/executor/join/hash_join_base.go:328]: why not put `syncerDone` as defer function?

File: pkg/executor/join/hash_join_probe_worker_v2.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+			if !ok {
+				return nil
+			}
+		}
+	}
+	if joinResult == nil {
+		return nil
+	} else if joinResult.err != nil || (joinResult.chk != nil && joinResult.chk.NumRows() > 0) {
+		w.HashJoinCtx.joinResultCh <- joinResult
+	} else if joinResult.chk != nil && joinResult.chk.NumRows() == 0 {
```
Comment (windtalker) [inline pkg/executor/join/hash_join_probe_worker_v2.go:64]: when will this happen?

File: pkg/executor/join/hash_join_v2.go [hunk 5]
```diff
@@ -401,184 +376,145 @@ func (e *HashJoinV2Exec) initializeForProbe() {
...
+		func(r any) {
+			handleError(e.joinResultCh, &e.finished, r)
+		},
+	)
 }
 
-func (e *HashJoinV2Exec) waitJoinWorkersAndCloseResultChan() {
-	e.workerWg.Wait()
+// finaWorker is responsible for scanning the row table after probe done and wake up the build fetcher
+func (e *HashJoinV2Exec) finalWorker(syncer chan struct{}) {
```
Comment (windtalker) [inline pkg/executor/join/hash_join_v2.go:456]: I think if the join don't need to scan hash table after probe, there is no need to start final worker?

File: pkg/executor/join/hash_join_v2.go [hunk 6]
```diff
@@ -588,40 +524,31 @@ func (w *ProbeWorkerV2) getNewJoinResult() (bool, *hashjoinWorkerResult) {
 func (e *HashJoinV2Exec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
 	if !e.prepared {
 		e.initHashTableContext()
+		e.initializeForProbe()
```
Comment (windtalker) [inline pkg/executor/join/hash_join_v2.go:527]: why move `e.initializeForProbe()` out from `e.fetchAndProbeHashTable()`?

File: pkg/executor/join/hash_join_v2.go [hunk 7]
```diff
@@ -588,40 +524,31 @@ func (w *ProbeWorkerV2) getNewJoinResult() (bool, *hashjoinWorkerResult) {
...
 	}
-	if e.ProbeSideTupleFetcher.shouldLimitProbeFetchSize() {
-		atomic.StoreInt64(&e.ProbeSideTupleFetcher.requiredRows, int64(req.RequiredRows()))
-	}
+
+	e.updateRequiredRows(req)
 	req.Reset()
 
 	result, ok := <-e.joinResultCh
+	e.recycleChunk(req, result)
```
Comment (windtalker) [inline pkg/executor/join/hash_join_v2.go:539]: why move this recycle code before error check?

File: pkg/executor/join/hash_join_v2.go [hunk 8]
```diff
@@ -677,166 +629,158 @@ func (e *HashJoinV2Exec) createTasks(buildTaskCh chan<- *buildTask, totalSegment
...
+	var err error
 
-	waitJobDone := func(wg *sync.WaitGroup, errCh chan error) bool {
-		wg.Wait()
-		close(errCh)
-		if err := <-errCh; err != nil {
-			e.buildFinished <- err
-			return false
+	for ifContinue {
+		ifContinue, err = e.dispatchBuildTasksImpl(syncer)
```
Comment (windtalker) [inline pkg/executor/join/hash_join_v2.go:657]: dispatchBuildTasksImpl can be called multiple times here?

File: pkg/executor/sortexec/sort_spill.go [hunk 9]
```diff
@@ -114,12 +114,13 @@ func (s *parallelSortSpillAction) actionImpl(t *memory.Tracker) bool {
 		s.spillHelper.cond.Wait()
 	}
 
+	// we may be in `needSpill` status, so it's necessary check if we are in `notSpilled` status
```
Comment (windtalker) [inline pkg/executor/sortexec/sort_spill.go:117]: unrelated changes?

File: pkg/executor/join/outer_join_probe.go [hunk 10]
```diff
@@ -110,37 +116,41 @@ func (j *outerJoinProbe) ScanRowTable(joinResult *hashjoinWorkerResult, sqlKille
...
 	remainCap := joinResult.chk.RequiredRows() - joinResult.chk.NumRows()
+
 	for insertedRows < remainCap && !j.rowIter.isEnd() {
-		currentRow := j.rowIter.getValue()
-		if !meta.isCurrentRowUsed(currentRow) {
-			// append build side of this row
-			j.appendBuildRowToCachedBuildRowsAndConstructBuildRowsIfNeeded(createMatchRowInfo(0, currentRow), joinResult.chk, 0, false)
-			insertedRows++
-		}
+		j.appendBuildRow(meta, joinResult.chk, j.rowIter.getValue(), &insertedRows)
```
Comment (windtalker) [inline pkg/executor/join/outer_join_probe.go:126]: why change this

File: pkg/executor/join/hash_join_v2.go [hunk 11]
```diff
@@ -677,166 +629,158 @@ func (e *HashJoinV2Exec) createTasks(buildTaskCh chan<- *buildTask, totalSegment
...
-	e.buildHashTable(buildTaskCh, wg, errCh, doneCh)
-	waitJobDone(wg, errCh)
-}
+	fetcherAndWorkerSyncer := &sync.WaitGroup{}
+
+	defer func() {
+		// We must ensure that all prebuild workers have exited before
+		// set `finished` flag and close buildFetcherAndDispatcherSyncChan
+		fetcherAndWorkerSyncer.Wait()
+		e.finished.Store(true)
```
Comment (windtalker) [inline pkg/executor/join/hash_join_v2.go:686]: It means this function need to wait all the probe finishes?

File: pkg/executor/join/hash_join_base.go (hunk 1 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_base.go:255]: > why not add `fetcherAndWorkerSyncer` as an argument of `fetchBuildSideRows` and let the caller to pass this argument

okk

File: pkg/executor/join/hash_join_base.go (hunk 2 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_base.go:258]: > You can define mulitple defers in a function, no need to use nested defer. For this specific case, you can move `defer close(chkCh)` to the first line of this function.

I will put this defer outside side of the defer function

File: pkg/executor/join/hash_join_probe_worker_v2.go (hunk 4 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_probe_worker_v2.go:64]: > when will this happen?

This logic is from the origin hash join.  Maybe this happens when there are no rows be successfully joined.

File: pkg/executor/join/hash_join_v2.go [hunk 12]
```diff
@@ -92,34 +103,43 @@ func (htc *hashTableContext) getCurrentRowSegment(workerID, partitionID int, tab
...
 
-func (htc *hashTableContext) finalizeCurrentSeg(workerID, partitionID int, builder *rowTableBuilder) {
-	seg := htc.getCurrentRowSegment(workerID, partitionID, nil, false, 0)
+func (htc *hashTableContext) finalizeCurrentSeg(workerID, partitionID int, builder *rowTableBuilder, needConsume bool) {
+	seg := htc.getCurrentRowSegment(workerID, partitionID, false, 0)
 	builder.rowNumberInCurrentRowTableSeg[partitionID] = 0
 	failpoint.Inject("finalizeCurrentSegPanic", nil)
 	seg.finalized = true
-	htc.memoryTracker.Consume(seg.totalUsedBytes())
+	if needConsume {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:111]: > when will `needConsume` be false?

When spill is triggered, `needConsume` will be false.

File: pkg/executor/join/hash_join_v2.go [hunk 13]
```diff
@@ -92,34 +103,43 @@ func (htc *hashTableContext) getCurrentRowSegment(workerID, partitionID int, tab
...
-	for i := 0; i < int(partitionNumber); i++ {
-		htc.hashTable.tables[i] = newSubTable(rowTables[i])
+
+	for i := 0; i < partitionNumber; i++ {
+		htc.hashTable.tables[i] = newSubTable(rowTables[i], nil)
 	}
-	htc.rowTables = nil
-	return totalSegmentCnt
+
+	htc.clearAllSegmentsInRowTable()
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:135]: > why keep `htc.rowTables`

Because `htc.rowTables` may be used in next restore round.

File: pkg/executor/join/hash_join_v2.go [hunk 14]
```diff
@@ -301,6 +299,7 @@ func (e *HashJoinV2Exec) Open(ctx context.Context) error {
 		e.prepared = false
 		return err
 	}
+	e.isMemoryClearedForTest = true
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:302]: > why set a test variable to true by default?

Each time finishing a restore round, we will check if all memory has been cleared. Only when there is memory not clear then this variable will be set to false. So we need to set it to true at the beginning.

File: pkg/executor/join/hash_join_v2.go (hunk 5 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:456]: > I think if the join don't need to scan hash table after probe, there is no need to start final worker?

Yes, but final worker will not create more go routines when there is no need to scan hash table.

File: pkg/executor/join/hash_join_v2.go (hunk 7 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:539]: > why move this recycle code before error check?

If error happens, we will directly return and fail to recycle chunk, this will let workers hang.

File: pkg/executor/join/hash_join_v2.go (hunk 8 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:657]: > dispatchBuildTasksImpl can be called multiple times here?

Yes, after spill is added, this function will be called multiple times/

File: pkg/executor/sortexec/sort_spill.go (hunk 9 above)
Comment (xzhangxian1008) [reply pkg/executor/sortexec/sort_spill.go:117]: > unrelated changes?

Yes, it's unrelated with hash join refine. But I find that we can refine the comment in sort spill, so I put it in this pr.

File: pkg/executor/join/outer_join_probe.go (hunk 10 above)
Comment (xzhangxian1008) [reply pkg/executor/join/outer_join_probe.go:126]: > why change this

I think it's more readable to put these codes into a function and reader could know the purpose of these codes from function name.

File: pkg/executor/join/hash_join_v2.go (hunk 11 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:686]: > It means this function need to wait all the probe finishes?

No, build fetcher could directly exit without waiting all probe finished.

File: pkg/executor/join/hash_join_v2.go (hunk 6 above)
Comment (xzhangxian1008) [reply pkg/executor/join/hash_join_v2.go:527]: > why move `e.initializeForProbe()` out from `e.fetchAndProbeHashTable()`?

Because `joinResultCh` in `e.initializeForProbe()` function needs to be initialized before starting build fetcher, as `joinResultCh` is also used for transmitting error.
